### PR TITLE
#1592 refund utility

### LINKED ISSUE #1592: Payment utility method: refund
## Is your feature request related to a problem? Please describe.

Need to be able to 'refund'/'credit'/'return' goods. Initially this will just be for prescriptions, but may extend to supplier credits.

## Describe the solution you'd like

A utility method, `refund(TransactionBatch[] batches)`

- Will take an array of `TransactionBatch`s - creates a `CustomerCredit` 
- Adds a matching but inverse `TransactionBatch` for each passed batch to the `CustomerCredit` created, with a type of `"stock_in"`

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #1592 

## Change summary

- Adds refund() utility
- Added a `createRecord` for a `refundLine` - there's really just one more method to write - `cancel`, and that should cover all our use cases. Then it will be very easy to see where the shared/duplicated code is and can revert a few things. Finding it much easier this way

## Testing

N/A

### Related areas to think about

N/A

### CODE REVIEW

File: src/utilities/modules/dispensary/refund.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * Utility to refund a collection of TransactionBatch records. Creates a
+ * single CustomerCredit and for each TransactionBatch that should be
+ * refund, a RefundLine record, which enters the batch back into stock
+ * and adds credit for the patient.
+ *
+ *
+ * @param {User}               currentUser The currently logged in user.
+ * @param {Name}               patient     The patient to refund to
+ * @param {TransactionBatch[]} batches     Array of batches to be refund
```
Comment (mark-prins) [inline src/utilities/modules/dispensary/refund.js]:
```suggestion
 * @param {TransactionBatch[]} batches     Array of batches to be refunded
```

File: src/utilities/modules/dispensary/refund.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+export const refund = (currentUser, patient, batches) => {
+  const forSamePatient = batches.every(({ transaction }) => {
+    const { otherPartyName } = transaction;
+    return patient.id === otherPartyName.id;
```
Comment (mark-prins) [inline src/utilities/modules/dispensary/refund.js]: `otherPartyName` couldn't be null, could it?

File: src/utilities/modules/dispensary/refund.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  });
+
+  if (!batches.length) throw new Error('Trying to refund void');
+  if (!forSamePatient) throw new Error('Batches for different patients!');
+
+  const sumOfCredit = batches.reduce((acc, { total }) => acc + total, 0);
```
Comment (mark-prins) [inline src/utilities/modules/dispensary/refund.js]: are the refunded transactions marked in some way, so they can't be refunded again?

Comment (joshxg) [reply src/utilities/modules/dispensary/refund.js]: No. Need to make another desktop issue to find out different/correct behaviour. A new field may need to be added to `[trans_line]` or something as you can just refund as many times as you like.

Can even refund the same item multiple times in the same credit!

image

File: src/utilities/modules/dispensary/refund.js (hunk 1 above)
Comment (joshxg) [reply src/utilities/modules/dispensary/refund.js]: cbe90b871c4bdc1e9864c3e184e393690062b99d

File: src/utilities/modules/dispensary/refund.js (hunk 2 above)
Comment (joshxg) [reply src/utilities/modules/dispensary/refund.js]: oops thats broken completely, cheers - e375cbae25c5ae62762dd3a0d19651c41e921df8

File: src/utilities/modules/dispensary/refund.js (hunk 3 above)
Comment (joshxg) [reply src/utilities/modules/dispensary/refund.js]: Hope you don't mind me addressing this elsewhere, possibly after a discussion!

Comment (mark-prins) [reply src/utilities/modules/dispensary/refund.js]: yes, I think it needs to be sorted separately - and especially because desktop needs a tidy up first!

**Pull request merged**